File: scripts/feed_writer.py
"""
Writes LLM-generated API insights as structured JSON documents to data/changes/
and maintains data/index.json as a chronological manifest.
"""

import json
import logging
import re
from datetime import date, datetime, timezone
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)

DATA_DIR = Path(__file__).parent.parent / "data"
CHANGES_DIR = DATA_DIR / "changes"
INDEX_PATH = DATA_DIR / "index.json"

INTERESTING_SCORE_THRESHOLD = 2


def _slugify(s: str) -> str:
    s = s.lower()
    s = re.sub(r"[^a-z0-9]+", "-", s)
    return s.strip("-")


def _make_slug(api: str, insight_date: str) -> str:
    """e.g. 2026-08-22-aiplatform-v1beta1"""
    return f"{insight_date}-{_slugify(api)}"


def _load_index() -> list[dict]:
    if INDEX_PATH.exists():
        try:
            return json.loads(INDEX_PATH.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            logger.warning("data/index.json is malformed — starting fresh")
    return []


def _save_index(entries: list[dict]) -> None:
    INDEX_PATH.write_text(
        json.dumps(entries, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )


def _display_path(path: Path) -> Path:
    try:
        return path.relative_to(Path.cwd())
    except ValueError:
        return path


def get_recent_feed_entries(
    api: str, insight_date: str, max_entries: int = 3
) -> tuple[Optional[str], list[dict]]:
    """Look up index.json to find today's entry and recent past entries for the given api.

    Returns:
        A tuple: (existing_today_content_string, list_of_past_entries)
    """
    index = _load_index()
    api_entries = [e for e in index if e.get("api") == api]
    if not api_entries:
        return None, []

    api_entries.sort(key=lambda e: (e.get("date", ""), e.get("slug", "")), reverse=True)

    existing_today_content = None
    recent_history: list[dict] = []

    for entry in api_entries:
        entry_date = entry.get("date")
        slug = entry.get("slug") or entry.get("id")
        if not slug or not entry_date:
            continue

        json_path = CHANGES_DIR / f"{slug}.json"
        if not json_path.exists():
            continue

        try:
            doc = json.loads(json_path.read_text(encoding="utf-8"))
            summary = doc.get("summary", "")
            details = doc.get("details", "")
            content_str = f"Summary: {summary}\n\nDetails: {details}"
        except Exception as e:
            logger.warning(f"Failed to read data file {json_path}: {e}")
            continue

        if entry_date == insight_date:
            existing_today_content = content_str
        else:
            if len(recent_history) < max_entries:
                recent_history.append(
                    {"date": entry_date, "slug": slug, "content": content_str}
                )

    return existing_today_content, recent_history


def write_insight(insight: dict, insight_date: Optional[str] = None) -> Optional[str]:
    """Write a single insight as structured JSON to data/changes/{slug}.json.

    Args:
        insight: Parsed LLM output dict.
        insight_date: ISO date string (YYYY-MM-DD). Defaults to today UTC.

    Returns:
        The slug of the written file, or None if skipped.
    """
    score = insight.get("interesting_score", 0)
    api = insight.get("api", "unknown")

    if score < INTERESTING_SCORE_THRESHOLD:
        logger.info(
            f"  Skipping {api} (interesting_score={score} < {INTERESTING_SCORE_THRESHOLD})"
        )
        return None

    if insight_date is None:
        insight_date = date.today().isoformat()

    CHANGES_DIR.mkdir(parents=True, exist_ok=True)

    base_slug = _make_slug(api, insight_date)
    file_path = CHANGES_DIR / f"{base_slug}.json"
    final_slug = base_slug

    service_name = insight.get("service_name") or insight.get("service", "Google Cloud")
    title = insight.get("title") or f"{service_name} API Update"
    summary = insight.get("summary", "")
    details = insight.get("details", summary)
    impact = str(insight.get("impact", "low")).lower()
    breaking = bool(insight.get("breaking", False))
    tags = insight.get("tags", [])
    if isinstance(tags, str):
        tags = [tags]

    # Extract RPC methods if present
    extracted_methods = insight.get("extracted_methods", [])
    if not extracted_methods:
        method_matches = re.findall(r"`([a-zA-Z0-9_]+(?:\.[a-zA-Z0-9_]+){2,})`", f"{summary} {details}")
        extracted_methods = list(dict.fromkeys(method_matches))[:6]

    entry = {
        "id": final_slug,
        "slug": final_slug,
        "date": insight_date,
        "api": api,
        "service": service_name,
        "service_name": service_name,
        "title": title,
        "summary": summary,
        "details": details,
        "impact": impact,
        "breaking": breaking,
        "interesting_score": score,
        "tags": tags,
        "extracted_methods": extracted_methods,
        "status": "canary",
        "radar_ring": "hold" if breaking else "assess",
        "lead_time_days": 14,
        "generated_at": datetime.now(timezone.utc).isoformat(),
    }

    is_update = file_path.exists()
    file_path.write_text(json.dumps(entry, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")

    if is_update:
        logger.info(f"  Updated: {_display_path(file_path)} (score={score})")
    else:
        logger.info(f"  Written: {_display_path(file_path)} (score={score})")

    # Update index
    index = _load_index()
    index_entry = {
        "id": final_slug,
        "slug": final_slug,
        "date": insight_date,
        "api": api,
        "service": service_name,
        "service_name": service_name,
        "title": title,
        "summary": summary,
        "impact": impact,
        "breaking": breaking,
        "interesting_score": score,
        "tags": tags,
        "extracted_methods": extracted_methods,
        "generated_at": entry["generated_at"],
    }

    # In case of update, replace existing entry in-place
    replaced = False
    for i, existing in enumerate(index):
        if existing.get("slug") == final_slug or existing.get("id") == final_slug:
            index[i] = index_entry
            replaced = True
            break

    if not replaced:
        index.append(index_entry)

    # Keep sorted newest-first
    index.sort(key=lambda e: (e["date"], e["slug"]), reverse=True)
    _save_index(index)

    return final_slug


def write_insights(
    insights: list[dict], insight_date: Optional[str] = None
) -> list[str]:
    """Write multiple insights to data/changes/.

    Returns:
        List of written slugs.
    """
    written = []
    for insight in insights:
        slug = write_insight(insight, insight_date)
        if slug:
            written.append(slug)
    return written

File: scripts/feed_writer.ts
/**
 * Writes LLM-generated API insights as structured JSON documents to data/changes/
 * and maintains data/index.json as a chronological manifest.
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { findServiceMeta } from './taxonomy.ts';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const DEFAULT_DATA_DIR = path.resolve(__dirname, '../data');
export const DEFAULT_CHANGES_DIR = path.join(DEFAULT_DATA_DIR, 'changes');
export const DEFAULT_INDEX_PATH = path.join(DEFAULT_DATA_DIR, 'index.json');

export const INTERESTING_SCORE_THRESHOLD = 3;

export function slugify(s: string): string {
  return s
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

export function makeSlug(api: string, insightDate: string): string {
  return `${insightDate}-${slugify(api)}`;
}

export function loadIndex(indexPath = DEFAULT_INDEX_PATH): any[] {
  if (fs.existsSync(indexPath)) {
    try {
      return JSON.parse(fs.readFileSync(indexPath, 'utf-8'));
    } catch {
      console.warn('data/index.json is malformed — starting fresh');
    }
  }
  return [];
}

export function saveIndex(entries: any[], indexPath = DEFAULT_INDEX_PATH): void {
  fs.writeFileSync(indexPath, JSON.stringify(entries, null, 2) + '\n', 'utf-8');
}

export function getRecentFeedEntries(
  api: string,
  insightDate: string,
  maxEntries = 3,
  dataDir = DEFAULT_DATA_DIR
): [string | null, any[]] {
  const indexPath = path.join(dataDir, 'index.json');
  const changesDir = path.join(dataDir, 'changes');

  const index = loadIndex(indexPath);
  const apiEntries = index.filter((e) => e.api === api);
  if (apiEntries.length === 0) {
    return [null, []];
  }

  apiEntries.sort((a, b) => {
    const da = `${a.date || ''}-${a.slug || a.id || ''}`;
    const db = `${b.date || ''}-${b.slug || b.id || ''}`;
    return db.localeCompare(da);
  });

  let existingTodayContent: string | null = null;
  const recentHistory: any[] = [];

  for (const entry of apiEntries) {
    const entryDate = entry.date;
    const slug = entry.slug || entry.id;
    if (!slug || !entryDate) continue;

    const jsonPath = path.join(changesDir, `${slug}.json`);
    if (!fs.existsSync(jsonPath)) continue;

    try {
      const doc = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));
      const summary = doc.summary || '';
      const details = doc.details || '';
      const contentStr = `Summary: ${summary}\n\nDetails: ${details}`;
      const extractedMethods = doc.extracted_methods || [];
      const targetPaths = doc.target_paths || doc.extracted_paths || doc.target_schema_paths || [];

      if (entryDate === insightDate) {
        existingTodayContent = contentStr;
      } else if (recentHistory.length < maxEntries) {
        recentHistory.push({
          date: entryDate,
          slug,
          content: contentStr,
          extracted_methods: extractedMethods,
          target_paths: targetPaths,
        });
      }
    } catch (err: any) {
      console.warn(`Failed to read data file ${jsonPath}:`, err.message);
    }
  }

  return [existingTodayContent, recentHistory];
}

export function writeInsight(
  insight: Record<string, any>,
  insightDate?: string,
  dataDir = DEFAULT_DATA_DIR
): string | null {
  const score = insight.interesting_score ?? 0;
  const api = insight.api || 'unknown';
  const isBreaking = Boolean(insight.breaking);

  // Breaking changes always get published regardless of score, because a dropped
  // breaking alert is worse than a noisy feed entry (insight 530514db).
  if (!isBreaking && score < INTERESTING_SCORE_THRESHOLD) {
    console.info(`  Skipping ${api} (interesting_score=${score} < ${INTERESTING_SCORE_THRESHOLD}, non-breaking)`);
    return null;
  }

  const resolvedDate = insightDate || new Date().toISOString().slice(0, 10);
  const changesDir = path.join(dataDir, 'changes');
  const indexPath = path.join(dataDir, 'index.json');

  fs.mkdirSync(changesDir, { recursive: true });

  const baseSlug = makeSlug(api, resolvedDate);
  const filePath = path.join(changesDir, `${baseSlug}.json`);
  const finalSlug = baseSlug;

  const apiId = api ? api.split('.')[0].toLowerCase() : '';
  const canonicalMeta = findServiceMeta(apiId) || findServiceMeta(insight.service_name || insight.service || '');
  const serviceName = canonicalMeta?.name || insight.service_name || insight.service || 'Google Cloud';
  const title = insight.title || `${serviceName} API Update`;
  const summary = insight.summary || '';
  const details = insight.details || summary;
  const impact = String(insight.impact || 'low').toLowerCase();
  const breaking = Boolean(insight.breaking);
  let tags = insight.tags || [];
  if (typeof tags === 'string') tags = [tags];

  let extractedMethods = insight.extracted_methods || [];
  if (extractedMethods.length === 0) {
    const combined = `${summary} ${details}`;
    const matches = combined.match(/`([a-zA-Z0-9_]+(?:\.[a-zA-Z0-9_]+){2,})`/g) || [];
    extractedMethods = Array.from(new Set(matches.map((m) => m.slice(1, -1)))).slice(0, 6);
  }

  const targetPaths = insight.target_paths || insight.extracted_paths || [];

  const entry = {
    id: finalSlug,
    date: resolvedDate,
    api,
    service: serviceName,
    title,
    summary,
    details,
    impact,
    breaking,
    interesting_score: score,
    tags,
    extracted_methods: extractedMethods,
    target_paths: targetPaths,
    status: 'canary',
    lead_time_days: null,
    generated_at: new Date().toISOString(),
  };

  const isUpdate = fs.existsSync(filePath);
  fs.writeFileSync(filePath, JSON.stringify(entry, null, 2) + '\n', 'utf-8');

  console.info(`  ${isUpdate ? 'Updated' : 'Written'}: ${filePath} (score=${score})`);

  // Update index
  const index = loadIndex(indexPath);
  const indexEntry = {
    id: finalSlug,
    date: resolvedDate,
    api,
    service: serviceName,
    title,
    summary,
    impact,
    breaking,
    interesting_score: score,
    tags,
    extracted_methods: extractedMethods,
    target_paths: targetPaths,
    generated_at: entry.generated_at,
  };

  let replaced = false;
  for (let i = 0; i < index.length; i++) {
    if (index[i].id === finalSlug || index[i].slug === finalSlug) {
      index[i] = indexEntry;
      replaced = true;
      break;
    }
  }
  if (!replaced) {
    index.push(indexEntry);
  }

  index.sort((a, b) => {
    const da = `${a.date || ''}-${a.id || a.slug || ''}`;
    const db = `${b.date || ''}-${b.id || b.slug || ''}`;
    return db.localeCompare(da);
  });
  saveIndex(index, indexPath);

  return finalSlug;
}

export function writeInsights(
  insights: Array<Record<string, any>>,
  insightDate?: string,
  dataDir = DEFAULT_DATA_DIR
): string[] {
  const written: string[] = [];
  for (const insight of insights) {
    const slug = writeInsight(insight, insightDate, dataDir);
    if (slug) {
      written.push(slug);
    }
  }
  return written;
}
